drivers/QSPI.cpp

```diff
@@ -0,0 +1,271 @@
+            if(true == _acquire()) {
+                qspi_command_t *qspi_cmd = _build_qspi_command(-1, address, -1);
+                if(QSPI_STATUS_OK == qspi_write(&_qspi, qspi_cmd, tx_buffer, tx_length)) {
+                    ret_status = QSPI_STATUS_OK; 
+                }
+            }
+            unlock();
+        }
+    } else {
+        ret_status = QSPI_STATUS_INVALID_PARAMETER; 
+    }
+    
+    return ret_status;
+}
+
+qspi_status_t QSPI::read(unsigned int instruction, unsigned int address, unsigned int alt, char *rx_buffer, size_t *rx_length) {
+    qspi_status_t ret_status = QSPI_STATUS_ERROR; 
+    
+    if( (rx_length != NULL) && (rx_buffer != NULL) ) {
+        if(*rx_length != 0) {
+            lock();
+            if( true == _acquire()) {
+                qspi_command_t *qspi_cmd = _build_qspi_command(instruction, address, alt);
+                if(QSPI_STATUS_OK == qspi_read(&_qspi, qspi_cmd, rx_buffer, rx_length)) {
+                    ret_status = QSPI_STATUS_OK; 
+                }
+            }
+            unlock();
+        }
+    } else {
+        ret_status = QSPI_STATUS_INVALID_PARAMETER; 
+    }
+    
+    return ret_status;
+}
+
+qspi_status_t QSPI::write(unsigned int instruction, unsigned int address, unsigned int alt, const char *tx_buffer, size_t *tx_length) {
+    qspi_status_t ret_status = QSPI_STATUS_ERROR; 
+    
+    if( (tx_length != NULL) && (tx_buffer != NULL) ) {
+        if(*tx_length != 0) {
+            lock();
+            if(true == _acquire()) {
+                qspi_command_t *qspi_cmd = _build_qspi_command(instruction, address, alt);
+                if(QSPI_STATUS_OK == qspi_write(&_qspi, qspi_cmd, tx_buffer, tx_length)) {
+                    ret_status = QSPI_STATUS_OK; 
+                }
+            }
+            unlock();
+        }
+    } else {
+        ret_status = QSPI_STATUS_INVALID_PARAMETER; 
+    }
+    
+    return ret_status;
+}
+
+qspi_status_t QSPI::command_transfer(unsigned int instruction, const char *tx_buffer, size_t tx_length, const char *rx_buffer, size_t rx_length) {
+    qspi_status_t ret_status = QSPI_STATUS_ERROR; 
+    
+    lock();
+    if(true == _acquire()) {
+        qspi_command_t *qspi_cmd = _build_qspi_command(instruction, -1, -1); //We just need the command
+        if(QSPI_STATUS_OK == qspi_command_transfer(&_qspi, qspi_cmd, (const void *)tx_buffer, tx_length, (void *)rx_buffer, rx_length)) {
+            //We got error status, return 0
+            ret_status = QSPI_STATUS_OK; 
+        }
+    } 
+    unlock();
+    
+    return ret_status;
+}
+
+void QSPI::lock() {
+    _mutex->lock();
+}
+
+void QSPI::unlock() {
+    _mutex->unlock();
+}
+
+// Note: Private function with no locking
+bool QSPI::_acquire() {
+    qspi_status_t ret = QSPI_STATUS_OK;
+    
+    if (_owner != this) {
+        //This will set freq as well
+        ret = qspi_init(&_qspi, _qspi_io0, _qspi_io1, _qspi_io2, _qspi_io3, _qspi_clk, _qspi_cs, _hz, _mode );
```

Comment:
Does `acquire` need to initialize ? If bus owner is being switched, what is required for peripheral to do? Only reconfigure freq ? Pins etc should be configured

Reply:
I think you can have different cs lines connected to different devices/flash mems and can also have different mode. That's why its more safe to re-init here.

Reply:
OK, we might need document this. Because if HAL keeps track what is ON, this init can easily fail. IF QSPI0 peripheral was already initialized, it would return with no action. If this is intention to allow this, we need to doc this in HAL what init should do (always allow to init peripheral). I'll work on docs, so should have this covered then.